File: app/src/touchy_pad/transport_net.py
"""TCP transport for the Touchy-Pad out-of-process simulator.

The simulator (``touchy simulator``) exposes the same protobuf
command/response protocol over a TCP socket that the firmware exposes
over USB bulk endpoints. Framing on the wire is *identical*: a 4-byte
little-endian length prefix followed by the nanopb payload (see
:mod:`touchy_pad.transport`).

This module provides:
  * :data:`DEFAULT_SIM_PORT` — the agreed-upon TCP port.
  * :data:`SIM_URL_ENV` — the env-var name used as a fallback for
    discovery (``TOUCHY_SIM_URL``).
  * :class:`TcpTransport` — a :class:`~touchy_pad.transport.Transport`
    that speaks the wire format above to a remote / in-process
    :class:`~touchy_pad.sim.server.SimServer`.
  * :func:`parse_sim_url` — split a ``host[:port]`` /
    ``tcp://host[:port]`` string into ``(host, port)``.

One TCP connection serves one command/response stream. Events are
still polled via ``EventConsumeCmd`` exactly as on the USB transport.
"""

from __future__ import annotations

import logging
import os
import socket
import threading
import time

from .transport import _LEN_STRUCT, _MAX_FRAME, Transport, TransportError, _pack

#: The default TCP port the out-of-process simulator listens on.
#: Exposed as a symbolic constant so the firmware/host code never
#: hardcodes the raw number twice.
DEFAULT_SIM_PORT: int = 8935

#: Environment variable consulted by :func:`touchy_pad.api.touchy_open`
#: (and the Rust client) before falling back to USB enumeration. Value
#: is parsed by :func:`parse_sim_url`.
SIM_URL_ENV: str = "TOUCHY_SIM_URL"

_log = logging.getLogger(__name__)


def parse_sim_url(value: str) -> tuple[str, int]:
    """Parse a ``host[:port]`` / ``tcp://host[:port]`` string.

    Returns ``(host, port)``; port defaults to :data:`DEFAULT_SIM_PORT`
    when omitted. IPv6 literals in square brackets are supported:
    ``[::1]:8935``.
    """
    s = value.strip()
    if s.startswith("tcp://"):
        s = s[len("tcp://") :]
    # Strip any trailing path (we don't use one, but be forgiving).
    s = s.split("/", 1)[0]

    host: str
    port = DEFAULT_SIM_PORT
    if s.startswith("["):
        end = s.index("]")
        host = s[1:end]
        rest = s[end + 1 :]
        if rest.startswith(":"):
            port = int(rest[1:])
    elif ":" in s:
        host, _, p = s.rpartition(":")
        port = int(p)
    else:
        host = s
    if not host:
        raise ValueError(f"empty host in sim URL: {value!r}")
    return host, port


def sim_url_from_env() -> tuple[str, int] | None:
    """Return ``(host, port)`` from :data:`SIM_URL_ENV`, or ``None``."""
    raw = os.environ.get(SIM_URL_ENV)
    if not raw:
        return None
    try:
        return parse_sim_url(raw)
    except (ValueError, IndexError) as exc:
        _log.warning("ignoring malformed %s=%r: %s", SIM_URL_ENV, raw, exc)
        return None


class TcpTransport(Transport):
    """Transport that talks the USB-bulk wire format over a TCP socket.

    Parameters
    ----------
    host:
        Hostname or IP address of the simulator server.
    port:
        TCP port (defaults to :data:`DEFAULT_SIM_PORT`).
    connect_timeout_ms:
        How long the constructor will retry ``socket.connect()`` before
        giving up. A short retry covers the "client races server" case
        common in tests.
    """

    # The TCP sim now consumes raw LVGL ``.bin`` exactly like the
    # firmware does (Stage 63), so the host-side image pipeline must
    # convert PNG/JPG/etc. to ``.bin`` before upload.
    needs_image_conversion = True

    def __init__(
        self,
        host: str = "127.0.0.1",
        port: int = DEFAULT_SIM_PORT,
        *,
        connect_timeout_ms: int = 5000,
    ) -> None:
        self._addr = (host, port)
        self._lock = threading.Lock()
        self._closed = False

        deadline = time.monotonic() + connect_timeout_ms / 1000.0
        last_exc: Exception | None = None
        while True:
            try:
                self._sock = socket.create_connection(self._addr, timeout=2.0)
                break
            except OSError as exc:
                last_exc = exc
                if time.monotonic() >= deadline:
                    raise TransportError(
                        f"could not connect to sim at {host}:{port}: {exc}"
                    ) from exc
                time.sleep(0.05)
        # Disable Nagle: command/response messages are small and
        # latency matters more than coalescing.
        try:
            self._sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        except OSError:
            pass
        _log.info("sim: TCP transport connected to %s:%d", host, port)
        del last_exc  # silence linter

    # -- Transport interface --------------------------------------------

    def send_command(self, payload: bytes) -> None:
        if self._closed:
            raise TransportError("TcpTransport is closed")
        frame = _pack(payload)
        with self._lock:
            try:
                self._sock.sendall(frame)
            except OSError as exc:
                self._closed = True
                raise TransportError(f"sim TCP send failed: {exc}") from exc

    def recv_response(self, timeout_ms: int = 2000) -> bytes:
        if self._closed:
            raise TransportError("TcpTransport is closed")
        self._sock.settimeout(timeout_ms / 1000.0)
        try:
            header = self._recv_exact(_LEN_STRUCT.size)
            (length,) = _LEN_STRUCT.unpack(header)
            if length > _MAX_FRAME:
                raise TransportError(f"sim TCP frame exceeds {_MAX_FRAME}-byte cap: {length} bytes")
            return self._recv_exact(length)
        except TimeoutError as exc:
            raise TransportError(f"sim TCP: no response within {timeout_ms} ms") from exc
        except OSError as exc:
            self._closed = True
            raise TransportError(f"sim TCP recv failed: {exc}") from exc

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        try:
            self._sock.shutdown(socket.SHUT_RDWR)
        except OSError:
            pass
        try:
            self._sock.close()
        except OSError:
            pass

    # -- helpers --------------------------------------------------------

    def _recv_exact(self, n: int) -> bytes:
        buf = bytearray()
        while len(buf) < n:
            chunk = self._sock.recv(n - len(buf))
            if not chunk:
                raise TransportError("sim TCP: connection closed by peer")
            buf.extend(chunk)
        return bytes(buf)


__all__ = [
    "DEFAULT_SIM_PORT",
    "SIM_URL_ENV",
    "TcpTransport",
    "parse_sim_url",
    "sim_url_from_env",
]

File: rust/touchy-pad/src/transport_net.rs
//! TCP transport for the out-of-process Python simulator.
//!
//! The simulator (`touchy simulator`) speaks the *same* length-prefixed
//! nanopb framing on TCP as the firmware does over USB bulk endpoints,
//! so this transport is the network counterpart of
//! [`crate::transport_usb::UsbTransport`].
//!
//! See `docs/simulator.md` and Stage 63 in `docs/design.md`.

use std::time::Duration;

use async_trait::async_trait;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::sync::Mutex;

use crate::error::{Result, TouchyError};
use crate::transport::{MAX_FRAME, Transport, pack};

/// Default TCP port the out-of-process simulator listens on. Mirrors
/// `touchy_pad.transport_net.DEFAULT_SIM_PORT`.
pub const DEFAULT_SIM_PORT: u16 = 8935;

/// Environment variable consulted by [`crate::Touchy::open`] before
/// USB enumeration. Same name as on the Python side.
pub const SIM_URL_ENV: &str = "TOUCHY_SIM_URL";

/// TCP-backed [`Transport`] talking to a `touchy simulator` process.
pub struct TcpTransport {
	inner: Mutex<TcpStream>,
	addr: String,
}

impl TcpTransport {
	/// Connect to `host:port`.
	pub async fn connect(host: &str, port: u16) -> Result<Self> {
		let addr = format!("{host}:{port}");
		let stream = TcpStream::connect(&addr).await.map_err(|e| TouchyError::Transport(format!("sim TCP connect {addr}: {e}")))?;
		stream.set_nodelay(true).ok();
		tracing_log(&format!("sim: TCP transport connected to {addr}"));
		Ok(Self { inner: Mutex::new(stream), addr })
	}

	/// Connect using a `host[:port]` / `tcp://host[:port]` URL string.
	pub async fn connect_url(url: &str) -> Result<Self> {
		let (host, port) = parse_sim_url(url)?;
		Self::connect(&host, port).await
	}
}

#[async_trait]
impl Transport for TcpTransport {
	async fn send_command(&self, payload: &[u8]) -> Result<()> {
		let frame = pack(payload)?;
		let mut g = self.inner.lock().await;
		g.write_all(&frame).await.map_err(|e| TouchyError::Transport(format!("sim TCP send to {}: {e}", self.addr)))?;
		Ok(())
	}

	async fn recv_response(&self, timeout: Duration) -> Result<Vec<u8>> {
		let mut g = self.inner.lock().await;
		let fut = async {
			let mut hdr = [0u8; 4];
			g.read_exact(&mut hdr).await.map_err(|e| TouchyError::Transport(format!("sim TCP recv header: {e}")))?;
			let len = u32::from_le_bytes(hdr) as usize;
			if len > MAX_FRAME {
				return Err(TouchyError::Framing(format!("sim TCP frame exceeds {MAX_FRAME}-byte cap: {len} bytes")));
			}
			let mut buf = vec![0u8; len];
			g.read_exact(&mut buf).await.map_err(|e| TouchyError::Transport(format!("sim TCP recv body: {e}")))?;
			Ok::<Vec<u8>, TouchyError>(buf)
		};
		match tokio::time::timeout(timeout, fut).await {
			Ok(r) => r,
			Err(_) => Err(TouchyError::Timeout(timeout)),
		}
	}

	fn needs_image_conversion(&self) -> bool {
		// The TCP sim consumes raw LVGL `.bin` exactly like the
		// firmware does, so the host pipeline must convert
		// PNG/JPG/etc. before upload.
		true
	}
}

/// Parse a `host[:port]` / `tcp://host[:port]` string.
///
/// Mirrors `touchy_pad.transport_net.parse_sim_url`. Supports IPv6
/// literals in square brackets (e.g. `[::1]:8935`). Returns
/// [`DEFAULT_SIM_PORT`] when no port is given.
pub fn parse_sim_url(value: &str) -> Result<(String, u16)> {
	let s = value.trim();
	let s = s.strip_prefix("tcp://").unwrap_or(s);
	let s = s.split('/').next().unwrap_or(s);

	let (host, port) = if let Some(rest) = s.strip_prefix('[') {
		// IPv6 literal
		let end = rest.find(']').ok_or_else(|| TouchyError::Other(format!("unterminated IPv6 literal in sim URL: {value:?}")))?;
		let host = &rest[..end];
		let after = &rest[end + 1..];
		let port = if let Some(p) = after.strip_prefix(':') {
			p.parse::<u16>().map_err(|e| TouchyError::Other(format!("bad port in sim URL {value:?}: {e}")))?
		} else {
			DEFAULT_SIM_PORT
		};
		(host.to_string(), port)
	} else if let Some(idx) = s.rfind(':') {
		let (h, p) = s.split_at(idx);
		let p = &p[1..];
		(h.to_string(), p.parse::<u16>().map_err(|e| TouchyError::Other(format!("bad port in sim URL {value:?}: {e}")))?)
	} else {
		(s.to_string(), DEFAULT_SIM_PORT)
	};

	if host.is_empty() {
		return Err(TouchyError::Other(format!("empty host in sim URL: {value:?}")));
	}
	Ok((host, port))
}

/// Return `Some((host, port))` from the [`SIM_URL_ENV`] env var, or
/// `None` if it is unset / malformed.
pub fn sim_url_from_env() -> Option<(String, u16)> {
	let raw = std::env::var(SIM_URL_ENV).ok()?;
	if raw.is_empty() {
		return None;
	}
	match parse_sim_url(&raw) {
		Ok(v) => Some(v),
		Err(e) => {
			tracing_log(&format!("ignoring malformed {SIM_URL_ENV}={raw:?}: {e}"));
			None
		}
	}
}

fn tracing_log(msg: &str) {
	// Library-side logging; callers wire in their own subscriber.
	log::info!("{msg}");
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn parse_host_port() {
		assert_eq!(parse_sim_url("127.0.0.1:8935").unwrap(), ("127.0.0.1".to_string(), 8935));
	}

	#[test]
	fn parse_host_only_default_port() {
		assert_eq!(parse_sim_url("example.test").unwrap(), ("example.test".to_string(), DEFAULT_SIM_PORT));
	}

	#[test]
	fn parse_tcp_prefix() {
		assert_eq!(parse_sim_url("tcp://10.0.0.1:1234").unwrap(), ("10.0.0.1".to_string(), 1234));
	}

	#[test]
	fn parse_ipv6() {
		assert_eq!(parse_sim_url("[::1]:9000").unwrap(), ("::1".to_string(), 9000));
		assert_eq!(parse_sim_url("[::1]").unwrap(), ("::1".to_string(), DEFAULT_SIM_PORT));
	}

	#[test]
	fn parse_rejects_empty_host() {
		assert!(parse_sim_url("").is_err());
		assert!(parse_sim_url(":8935").is_err());
	}
}
